### PR TITLE
Feature:  AES CTR encryption

### PR DESCRIPTION
**Pull request by matthewvon**
This PR has two major goals:

1. Make rocksdb's existing env_encryption.{h,cc} fit normal C++ pattern of declarations in .h and definitions in .cc so that new encryption code can inherit from existing work.

2. Implement AES CTR encryption in a manner that assumes users and programmers will need various types of migration of time:  code migration, migration to / from encryption, and migration of underlying encryption keys.

### CODE REVIEW

**Comment by matthewvon** [comment]
FYI:  Goal #1 to make EncrytpedEnv inheritable has been accepted and merged by Facebook.

File: include/rocksdb/env_encrypt2.h [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+      memcpy(desc, Desc, DescLen);
+      valid = true;
+    }
+  }
+
+  Sha1Description_t(const std::string& key_desc_str);
+
+  // goal is to explicitly remove desc from memory once no longer needed
+  ~Sha1Description_t() {
+    memset(desc, 0, EVP_MAX_MD_SIZE);
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply include/rocksdb/env_encrypt2.h:57]: There is a security guideline somewhere about always explicitly erase in-memory keys.  Probably in a really old book on my shelf.   The Sha1Description erase is not really necessary, but the code is then consistent with AesCtrKey code where it is highly desired.

File: include/rocksdb/env_encrypt2.h [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+                                 sizeof(prefix_buffer.key_description_));
+
+          auto it = encrypt_read_.find(desc);
+          if (encrypt_read_.end() != it) {
+            CTREncryptionProvider2* ptr =
+                (CTREncryptionProvider2*)it->second.get();
+            provider = it->second;
+            stream.reset(new AESBlockAccessCipherStream(
+                ptr->key(), code_version, prefix_buffer.nonce_));
+            ;
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply include/rocksdb/env_encrypt2.h:310]: bug.  will fix

File: include/rocksdb/env_encrypt2.h [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+    } else {
+      valid = false;
+    }
+  }
+
+  AesCtrKey_t(const std::string& key_str);
+
+  // goal is to explicitly remove key from memory once no longer needed
+  ~AesCtrKey_t() {
+    memset(key, 0, EVP_MAX_KEY_LENGTH);
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply include/rocksdb/env_encrypt2.h:92]: See previous comment about Sha1Description destructor.  This one is highly desired for in-memory keys.  Sha1Description's destructor is solely about having it look consistent with this destructor.  Also, theses destructors are not used very often currently.  Like once on shutdown.

[the remark this answers is not in the document]
Comment (toktarev) [reply include/rocksdb/env_encrypt2.h:92]: ok

File: include/rocksdb/env_encrypt2.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (toktarev) [reply include/rocksdb/env_encrypt2.h:57]: ok

File: include/rocksdb/env_encrypt2.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthewvon) [reply include/rocksdb/env_encrypt2.h]: @toktarev checked in.  along with additional comments about memset for first two destructors.

**Comment by matthewvon** [comment]
@buzztroll I have executed the rocksdb "make check".  It finished without errors.  I am comfortable squash merging.  I believe stardog/WORKSPACE pulls from a tag and will not be immediately impacted.

**Pull request merged**